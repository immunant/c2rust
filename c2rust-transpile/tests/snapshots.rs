use std::env::current_dir;
use std::fs;
use std::path::Path;
use std::process::Command;

use c2rust_transpile::{ReplaceMode, TranspilerConfig};

fn config() -> TranspilerConfig {
    TranspilerConfig {
        dump_untyped_context: false,
        dump_typed_context: false,
        pretty_typed_context: false,
        dump_function_cfgs: false,
        json_function_cfgs: false,
        dump_cfg_liveness: false,
        dump_structures: false,
        verbose: false,
        debug_ast_exporter: false,
        incremental_relooper: true,
        fail_on_multiple: false,
        filter: None,
        debug_relooper_labels: false,
        prefix_function_names: None,
        translate_asm: true,
        use_c_loop_info: true,
        use_c_multiple_info: true,
        simplify_structures: true,
        panic_on_translator_failure: false,
        emit_modules: false,
        fail_on_error: true,
        replace_unsupported_decls: ReplaceMode::Extern,
        translate_valist: true,
        overwrite_existing: true,
        reduce_type_annotations: false,
        reorganize_definitions: false,
        enabled_warnings: Default::default(),
        emit_no_std: false,
        output_dir: None,
        translate_const_macros: false,
        translate_fn_macros: false,
        disable_refactoring: false,
        preserve_unused_functions: false,
        log_level: log::LevelFilter::Warn,
        emit_build_files: false,
        binaries: Vec::new(),
    }
}

fn transpile(platform: Option<&str>, c_path: &Path) {
    let status = Command::new("clang")
        .args(&["-c", "-o", "/dev/null"])
        .arg(c_path)
        .status();
    assert!(status.unwrap().success());

    let (_temp_dir, temp_path) =
        c2rust_transpile::create_temp_compile_commands(&[c_path.to_owned()]);
    c2rust_transpile::transpile(config(), &temp_path, &[]);
    let cwd = current_dir().unwrap();
    let c_path = c_path.strip_prefix(&cwd).unwrap();
    let rs_path = c_path.with_extension("rs");
    let rs = fs::read_to_string(&rs_path).unwrap();
    let debug_expr = format!("cat {}", rs_path.display());

    let name = platform
        .map(|platform| ["transpile", platform].join("-"))
        .unwrap_or("transpile".into());

    insta::assert_snapshot!(name, &rs, &debug_expr);

    let status = Command::new("rustc")
        .args(&["--crate-type", "lib", "--edition", "2021", "-o", "-"])
        .arg(&rs_path)
        .status();
    assert!(status.unwrap().success());
}

#[test]
fn transpile_all() {
    insta::glob!("snapshots/*.c", |x| transpile(None, x));

    // Some things transpile differently on Linux vs. macOS,
    // as they use `unsigned long` and `unsigned long long` differently for builtins.
    // This makes snapshot tests trickier, as the output will be OS-dependent.
    // We handle this by adding OS name to the snapshot result filename.
    #[allow(unused)]
    let platform = "unknown";

    #[cfg(target_os = "linux")]
    let platform = "linux";
    #[cfg(target_os = "macos")]
    let platform = "macos";

    insta::with_settings!({snapshot_suffix => platform}, {
        insta::glob!("snapshots/platform-specific/*.c", |x| transpile(Some(platform), x));
    });
}
