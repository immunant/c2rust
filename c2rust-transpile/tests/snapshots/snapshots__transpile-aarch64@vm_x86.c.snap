---
source: c2rust-transpile/tests/snapshots.rs
expression: cat tests/snapshots/arch-specific/vm_x86.x86_64.rs
input_file: c2rust-transpile/tests/snapshots/arch-specific/vm_x86.c
---
#![allow(
    dead_code,
    non_camel_case_types,
    non_snake_case,
    non_upper_case_globals,
    unused_assignments,
    unused_mut
)]
#![feature(raw_ref_op)]
#[derive(Copy, Clone)]
#[repr(C)]
pub struct vm_t {
    pub programStack: ::core::ffi::c_int,
    pub entryOfs: ::core::ffi::c_int,
    pub dataBase: *mut ::core::ffi::c_void,
    pub codeBase: *mut ::core::ffi::c_void,
    pub instructionPointers: *mut ::core::ffi::c_ulong,
}
pub type byte = ::core::ffi::c_uchar;
pub const MAX_VMMAIN_ARGS: ::core::ffi::c_int = 50 as ::core::ffi::c_int;
#[no_mangle]
pub unsafe extern "C" fn VM_CallCompiled(
    mut vm: *mut vm_t,
    mut args: *mut ::core::ffi::c_int,
) -> ::core::ffi::c_int {
    let mut stack: [byte; 271] = [0; 271];
    let mut entryPoint: *mut ::core::ffi::c_void = 0 as *mut ::core::ffi::c_void;
    let mut programStack: ::core::ffi::c_int = 0;
    let mut stackOnEntry: ::core::ffi::c_int = 0;
    let mut image: *mut byte = 0 as *mut byte;
    let mut opStack: *mut ::core::ffi::c_int = 0 as *mut ::core::ffi::c_int;
    let mut opStackOfs: ::core::ffi::c_int = 0;
    let mut arg: ::core::ffi::c_int = 0;
    let mut currentVM: *mut vm_t = vm;
    stackOnEntry = (*vm).programStack;
    programStack = stackOnEntry;
    image = (*vm).dataBase as *mut byte;
    programStack -= 8 as ::core::ffi::c_int + 4 as ::core::ffi::c_int * MAX_VMMAIN_ARGS;
    arg = 0 as ::core::ffi::c_int;
    while arg < MAX_VMMAIN_ARGS {
        *(&raw mut *image.offset(
            (programStack + 8 as ::core::ffi::c_int + arg * 4 as ::core::ffi::c_int) as isize,
        ) as *mut ::core::ffi::c_int) = *args.offset(arg as isize);
        arg += 1;
    }
    *(&raw mut *image.offset((programStack + 4 as ::core::ffi::c_int) as isize)
        as *mut ::core::ffi::c_int) = 0 as ::core::ffi::c_int;
    *(&raw mut *image.offset(programStack as isize) as *mut ::core::ffi::c_int) =
        -(1 as ::core::ffi::c_int);
    entryPoint = ((*vm).codeBase).offset((*vm).entryOfs as isize);
    opStack = (&raw mut stack as *mut byte as *mut ::core::ffi::c_int)
        .offset(16 as ::core::ffi::c_int as isize);
    *opStack = 0 as ::core::ffi::c_int;
    opStackOfs = 0 as ::core::ffi::c_int;
    if opStackOfs != 1 as ::core::ffi::c_int
        || *opStack as ::core::ffi::c_uint != 0xdeadbeef as ::core::ffi::c_uint
    {
        return 0 as ::core::ffi::c_int;
    }
    if programStack
        != stackOnEntry - (8 as ::core::ffi::c_int + 4 as ::core::ffi::c_int * MAX_VMMAIN_ARGS)
    {
        return 0 as ::core::ffi::c_int;
    }
    (*vm).programStack = stackOnEntry;
    return *opStack.offset(opStackOfs as isize);
}
